File: widgetssdk/src/test/java/com/glia/widgets/filepreview/domain/GetImageFileFromDownloadsUseCaseTest.java
package com.glia.widgets.filepreview.domain;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import android.graphics.Bitmap;

import com.glia.widgets.filepreview.data.GliaFileRepository;
import com.glia.widgets.filepreview.domain.exception.FileNameMissingException;
import com.glia.widgets.filepreview.domain.usecase.GetImageFileFromDownloadsUseCase;

import org.junit.Before;
import org.junit.Test;

import io.reactivex.Maybe;

public class GetImageFileFromDownloadsUseCaseTest {

    private GliaFileRepository fileRepository;
    private GetImageFileFromDownloadsUseCase useCase;

    @Before
    public void setUp() {
        fileRepository = mock(GliaFileRepository.class);
        useCase = new GetImageFileFromDownloadsUseCase(fileRepository);
    }

    @Test
    public void execute_returnOnError_whenFileNameIsNull() {
        useCase.execute(null)
                .test()
                .assertError(FileNameMissingException.class);
    }

    @Test
    public void execute_returnOnError_whenFileNameIsEmpty() {
        useCase.execute(NAME_EMPTY)
                .test()
                .assertError(FileNameMissingException.class);
    }

    @Test
    public void execute_returnOnComplete() {
        when(fileRepository.loadImageFromDownloads(any()))
                .thenReturn(Maybe.just(BITMAP));
        useCase.execute(NAME)
                .test()
                .assertResult(BITMAP);
    }

    private static final String NAME = "NAME";
    private static final String NAME_EMPTY = "";
    private static final Bitmap BITMAP = mock(Bitmap.class);
}

File: widgetssdk/src/test/java/com/glia/widgets/filepreview/domain/GetImageFileFromDownloadsUseCaseTest.kt
package com.glia.widgets.filepreview.domain

import android.graphics.Bitmap
import com.glia.widgets.filepreview.data.GliaFileRepository
import com.glia.widgets.filepreview.domain.exception.FileNameMissingException
import com.glia.widgets.filepreview.domain.usecase.GetImageFileFromDownloadsUseCase
import io.reactivex.Maybe
import org.junit.Before
import org.junit.Test
import org.mockito.kotlin.any
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever

class GetImageFileFromDownloadsUseCaseTest {
    private lateinit var fileRepository: GliaFileRepository
    private lateinit var useCase: GetImageFileFromDownloadsUseCase

    @Before
    fun setUp() {
        fileRepository = mock()
        useCase = GetImageFileFromDownloadsUseCase(fileRepository)
    }

    @Test
    fun execute_emitFileNameMissingException_whenFileNameIsNull() {
        useCase(null)
            .test()
            .assertError(FileNameMissingException::class.java)
    }

    @Test
    fun execute_emitFileNameMissingException_whenFileNameIsEmpty() {
        useCase(NAME_EMPTY)
            .test()
            .assertError(FileNameMissingException::class.java)
    }

    @Test
    fun execute_successfullyCompletes_whenValidArgument() {
        whenever(fileRepository.loadImageFromDownloads(any())) doReturn Maybe.just(BITMAP)
        useCase(NAME)
            .test()
            .assertResult(BITMAP)
    }

    companion object {
        private const val NAME = "NAME"
        private const val NAME_EMPTY = ""
        private val BITMAP: Bitmap = mock()
    }
}
